src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java

```diff
@@ -101,63 +117,104 @@ public void initialise() {
+        HitResult hitResult = physics.rayTrace(start, direction, distance, Sets.newHashSet(character), PHYSICSFILTER);
+        if (!hitResult.isHit() || !hitResult.isWorldHit()) {
+            return Optional.empty();
+        }
+        Vector3i pos = new Vector3i(hitResult.getBlockPosition());
+        pos.add(Side.inDirection(hitResult.getHitNormal()).direction());
+        if (worldProvider.getBlock(pos).equals(air)) {
+            return Optional.of(pos);
+        } else {
+            return Optional.empty();
         }
-        return Optional.empty();
     }
 
     /**
      * Fill up the provided fluid container item with the current fluid interacted with in the game world.
-     *
-     * @param event          The event which has details about how this entity was activated.
-     * @param item           The reference to the item being activated.
-     * @param fluidContainer The component used for storing fluid in a container.
-     * @param itemComponent  A component included for filtering out non-matching events. Here, we only want entities
-     *                       which are used as items.
      */
+    @ReceiveEvent(priority = EventPriority.PRIORITY_HIGH)
+    public void fillFluidContainerItem(AttackRequest event, EntityRef character, CharacterComponent characterComponent) {
+        EntityRef item = event.getItem();
+        FluidContainerItemComponent fluidContainer = item.getComponent(FluidContainerItemComponent.class);
+        if (fluidContainer != null && (fluidContainer.fluidType == null || fluidContainer.volume < fluidContainer.maxVolume)) {
+
+            OnItemUseEvent onItemUseEvent = new OnItemUseEvent();
+            character.send(onItemUseEvent);
+            if (!onItemUseEvent.isConsumed()) {
+                EntityRef gaze = GazeAuthoritySystem.getGazeEntityForCharacter(character);
+                LocationComponent gazeLocation = gaze.getComponent(LocationComponent.class);
+                getLiquidInReach(JomlUtil.from(gazeLocation.getWorldPosition()), JomlUtil.from(gazeLocation.getWorldDirection()), character, characterComponent.interactionRange).ifPresent(pos -> {
+                    String fluidType = fluidRegistry.getCorrespondingFluid(worldProvider.getBlock(pos));
+                    if (fluidType != null && (fluidContainer.fluidType == null || fluidContainer.fluidType.equals(fluidType))) {
+                        EntityRef owner = item.getOwner();
+                        final EntityRef removedItem = inventoryManager.removeItem(owner, event.getInstigator(), item, false, 1);
+                        if (removedItem != null) {
+                            float blockAmount = getLiquidInBlock(pos);
+
+                            FluidContainerItemComponent fluidComponent = removedItem.getComponent(FluidContainerItemComponent.class);
+                            float totalAmount = blockAmount + fluidComponent.volume;
+                            if (totalAmount > fluidComponent.maxVolume) {
```

Comment:
For my taste that's a few too many nesting levels... can you flatten that a bit or pull some of it out into an extra function?